### ISSUE TITLE
[doc] how to prevent pytorch-nightly from being replaced by a released version on pip install

### ISSUE THREAD

**Issue by stas00** [pytorch/pytorch#49445]
## 📚 Documentation

I found another issue with pip and pytorch-nightly. 

If a user installs pytorch-nightly using `pip install --pre` which allows for pre-released versions as prescribed on https://pytorch.org/get-started/locally/, e.g.:
```
pip install --pre torch torchvision -f https://download.pytorch.org/whl/nightly/cu110/torch_nightly.html -U
```

installing any other package that includes `torch` in its requirements will wipe out the nightly build and install the latest release instead. (not 100% sure if it's with any packages depending on pytorch, but definitely seems to be the case with any package needing to do a binary cuda extension build).

So any such subsequent `pip install` commands for packages depending on pytorch, need to continue passing `--pre` to keep the preleased version installed - I have no idea where this could be documented.

**Comment by ezyang** [comment]
@stas00 Looking here:

fairscale deps: https://github.com/facebookresearch/fairscale/blob/master/requirements.txt
fairseq deps: https://github.com/pytorch/fairseq/blob/master/setup.py#L193

I see fairscale specifies a version bound, but fairseq does not. Could that be the problem? 

By the wording of https://www.python.org/dev/peps/pep-0440/#handling-of-pre-releases pip is supposed to accept prereleases if they are already installed, but I could imagine a bug where pip is accidentally filtering out prereleases unconditionally. Knowing the pip version would help too.

**Comment by stas00** [comment]
I tested your hypotheses by replacing " torch >= 1.4.0" with just "torch" in `fairscale`'s `requirements.txt` - I don't see any change - it still fetches pt-1.7.1

> Knowing the pip version would help too.

The very latest: 20.3.3

Here is how I understand the issue: the new pip build isolation feature (default!) requires fetching and installing all the project dependencies in a dedicated /tmp/pip-install-zyz tree - then it builds the package in question. So it goes and searches for the latest pytorch, which is pt-1.7.1 as of this writing and uses it. This is it. Since we didn't give it the repo with the nightly builds it doesn't find them and it doesn't care what has been already installed.

This (isolation build) is pretty much what conda does except it performs the same thing during conda package build rather than having this happen during user install.

**Comment by stas00** [comment]
The other workaround is to not use pip during the build of the cuda extension package:
```
git clone https://github.com/facebookresearch/fairscale/
cd fairscale
python setup.py bdist_wheel
```
and then install the wheel:
```
pip install dist/fairscale-0.1.1-cp38-cp38-linux_x86_64.whl
```

**Comment by ezyang** [comment]
Oh!! Yes it seems like isolation build would prevent a preinstalled prerelease from being used. Ugh, that doesn't seem great. I wonder if there's a pip bug about this.

**Comment by stas00** [comment]
I had a brief look under https://github.com/pypa/pip/issues searching for either `ninja` or `isolation` and didn't find anything pertinent, but it doesn't mean it's not there.

**Comment by rgommers** [comment]
> Yes it seems like isolation build would prevent a preinstalled prerelease from being used. Ugh, that doesn't seem great. I wonder if there's a pip bug about this.

Indeed.

There was a ton of discussion on this. Build isolation is enabled by default for projects that use a `pyproject.toml` (which both `fairscale` and `fairseq` have), and that's unlikely to change. The `pip` devs got a ton of bug reports about bugs that were due to non-reproducible builds, so are very invested in making wheel builds reproducible by default.

You can simply pass `--no-build-isolation` to `pip install` to avoid the problem.

**Comment by rgommers** [comment]
I'm not sure if there's a good place to document this. The best place would be to make it more obvious what's happening in the `pip` command line output. Currently this looks like:
```
$ pip install fairscale
Collecting fairscale
  Downloading fairscale-0.3.7.tar.gz (183 kB)
     |████████████████████████████████| 183 kB 4.4 MB/s 
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
    Preparing wheel metadata ... done
Collecting torch>=1.6.0
  Downloading torch-1.8.1-cp38-cp38-manylinux1_x86_64.whl (804.1 MB)
     |████████████████████████████████| 804.1 MB 3.1 kB/s 
Collecting typing-extensions
  Downloading typing_extensions-3.10.0.0-py3-none-any.whl (26 kB)
Collecting numpy
  Downloading numpy-1.20.3-cp38-cp38-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (15.4 MB)
     |████████████████████████████████| 15.4 MB 54.1 MB/s 
Building wheels for collected packages: fairscale
  Building wheel for fairscale (PEP 517) ... done
  Created wheel for fairscale: filename=fairscale-0.3.7-py3-none-any.whl size=231175 sha256=bdf8b5a6935f6ad406e88d0e7044c85eb3c34e5d71958d44104466d6121d96ed
  Stored in directory: /home/rgommers/.cache/pip/wheels/6a/6b/d8/a6f9094c018df2bfc55d0284c35d76b1b6b392edabbc0ed43a
Successfully built fairscale
Installing collected packages: typing-extensions, numpy, torch, fairscale
Successfully installed fairscale-0.3.7 numpy-1.20.3 torch-1.8.1 typing-extensions-3.10.0.0
```

The key line is:
```
Building wheel for fairscale (PEP 517)
```

PEP 517 is the thing that enables build isolation in the presence of `pyproject.toml`. Of course for users not aware of build isolation (which is, the vast majority of users) the phrase `PEP 517` is meaningless. 

Would it have helped to change this to `with build isolation` or `in an isolated environment`?

**Comment by stas00** [comment]
That would be a definite improvement to replace `PEP 517`, but most users won't know what build isolation is either.

And it's not like a user gets to choose whether to keep the nightly or not. So it's very unlikely that clarifying that one log line will bring this closer to the solution.

I think the problem also lies in pip, since if I use:
```
pip install fairscale --pre --f https://download.pytorch.org/whl/nightly/cu110/torch_nightly.html
```
it works too. As in the absence of `--pre --f https://download.pytorch.org/whl/nightly/cu110/torch_nightly.html` it's not aware that it has a non-standard released version of torch installed. So ideally, it should have some sort of flag saved on the original install of nightly. And when other packages get installed it should consult this flag and not fetch another torch unless other packages explicitly call for `torch==x.y.z`. So in the case of build isolation it should copy the local version into the isolation and not fetch an external one.
Or to remember where the external one was fetched from (not always possible, e.g. if it was installed from a local wheel).

One more possibility for pip: before starting a build in isolation, check if the installed dependencies match the ones that are about to be fetched and if there is a mismatch let user know?

So something to discuss in the pip land perhaps?
